src/astarte/handler.rs

```diff
@@ -116,26 +118,62 @@ impl DeviceSubscriber {
         }
     }
 
-    async fn on_event(
-        &mut self,
-        event: Result<DeviceEvent, AstarteError>,
-    ) -> Result<(), DeviceError> {
-        let event = event.map_err(DeviceError::Astarte)?;
+    async fn on_event(&mut self, event: Result<DeviceEvent, RecvError>) -> Result<(), DeviceError> {
+        // determine if sending a proto message hub error or a correct message to the message hub nodes
+        let (msghub_event, to_notify) = match event {
+            Ok(event) => {
+                let iface_name = event.interface.clone();
 
-        let itf_name = event.interface.clone();
+                let msg = MessageHubEvent::from(AstarteMessage::from(event));
 
-        let msg = AstarteMessage::from(event);
+                (msg, NotifySubscribers::Some(iface_name))
+            }
+            Err(err) => {
+                let msg = MessageHubEvent::from_error(&err);
+
+                if let RecvError::MappingNotFound { interface, mapping } = err {
+                    error!("mapping {mapping} not found");
+                    let iface_name = Interface::from_str(&interface)?
+                        .interface_name()
+                        .to_string();
+                    (msg, NotifySubscribers::Some(iface_name))
+                } else {
+                    (msg, NotifySubscribers::All)
+                }
+            }
+        };
 
-        let sub = self.subscribers.read().await;
-        let subscribers = sub
-            .values()
-            .filter(|subscriber| subscriber.introspection.contains(&itf_name));
+        trace!("message hub event: {msghub_event:?}");
+
+        match to_notify {
+            // TODO: since the error doesn't bring with itself interface information, we send the
+            //      event to all the nodes. This is tracked with SDK Issue #[358]
+            NotifySubscribers::All => self.send_to_subscribers(msghub_event, |_| true).await,
+            NotifySubscribers::Some(iface) => {
+                self.send_to_subscribers(msghub_event, |s| s.introspection.contains(&iface))
+                    .await
+            }
+        }
+    }
 
-        for subscriber in subscribers {
+    /// Send a message to a subset of subscribers.
+    ///
+    /// The closure passed in input to the method discriminates which subscribers must receive a message.
+    async fn send_to_subscribers<F>(
+        &self,
+        msghub_event: MessageHubEvent,
+        f: F,
+    ) -> Result<(), DeviceError>
+    where
+        F: Fn(&&Subscriber) -> bool,
+    {
+        let subscribers = self.subscribers.read().await;
+
+        for subscriber in subscribers.values().filter(f) {
```

Comment:
This seems complicated since in one of the cases you pass a function always returning `true`.

If you want to keep the function you could pass an iterator to it.